### PR TITLE
C# local functions

### PR DESCRIPTION
**Pull request by rpetrusha**
# C# local functions

The documentation for C# local functions, along with a revision of CS0106 to include errors generated for local functions.

## Suggested Reviewers

@BillWagner @omarmaher100

### CODE REVIEW

File: docs/csharp/language-reference/compiler-messages/cs0106.md [hunk 1]
```diff
@@ -36,11 +36,15 @@ The modifier 'modifier' is not valid for this item
  A class or interface member was marked with an invalid access modifier. The following examples describe some of these invalid modifiers:  
   
 -   The [static](../../../csharp/language-reference/keywords/static.md) and [public](../../../csharp/language-reference/keywords/public.md) modifiers are not permitted on interface methods.  
-  
+
+- The [static](../../../csharp/language-reference/keywords/static.md) modifier is not permitted on a local function. Whether a local function is static or instance is defined by its containing member.
```
Comment (svick) [inline docs/csharp/language-reference/compiler-messages/cs0106.md]: > Whether a local function is static or instance is defined by its containing member.

If a local function inside an instance method does not close over `this`, then it's generated as a `static` method. What does actually define whether a local function is static or instance? Whether it *could* access `this`? Or what IL is generated for it?

So, maybe just remove this sentence altogether? I'm not sure it's actually useful.

File: docs/csharp/language-reference/compiler-messages/cs0106.md [hunk 2]
```diff
@@ -36,11 +36,15 @@ The modifier 'modifier' is not valid for this item
...
+   
 -   The `public` keyword is not allowed on an explicit interface declaration. In this case, remove the `public` keyword from the explicit interface declaration.  
   
 -   The [abstract](../../../csharp/language-reference/keywords/abstract.md) keyword is not allowed on an explicit interface declaration because an explicit interface implementation can never be overridden.  
   
+- The [private](../keywords/private.md) modifier is not allowed on a local function. Local functions are always private.
```
Comment (svick) [inline docs/csharp/language-reference/compiler-messages/cs0106.md]: Why single out `private` here? No access modifiers are allowed on a local function.

File: docs/csharp/language-reference/compiler-messages/cs0106.md [hunk 3]
```diff
@@ -36,11 +36,15 @@ The modifier 'modifier' is not valid for this item
  A class or interface member was marked with an invalid access modifier. The following examples describe some of these invalid modifiers:  
   
 -   The [static](../../../csharp/language-reference/keywords/static.md) and [public](../../../csharp/language-reference/keywords/public.md) modifiers are not permitted on interface methods.  
-  
+
+- The [static](../../../csharp/language-reference/keywords/static.md) modifier is not permitted on a local function. Whether a local function is static or instance is defined by its containing member.
+   
 -   The `public` keyword is not allowed on an explicit interface declaration. In this case, remove the `public` keyword from the explicit interface declaration.  
   
 -   The [abstract](../../../csharp/language-reference/keywords/abstract.md) keyword is not allowed on an explicit interface declaration because an explicit interface implementation can never be overridden.
```
Comment (svick) [inline docs/csharp/language-reference/compiler-messages/cs0106.md]: Local functions also can't be `virtual`, `abstract`, `extern` or `partial`. Trying to do that does not currently cause CS0106, so this probably shouldn't be mentioned here. But I think it should do so, so I have filed https://github.com/dotnet/roslyn/issues/20279.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+- Event accessors
+- Anonymous methods
+- Lambda expressions
+- Finalizers
+- Other local functions
+
+However, local functions can't be declared inside an expression-bodied member.
+
+> [!NOTE]
+> In some cases, you can use a lambda expression to implement functionality also supported by a lambda expression. For a comparison, see [Local functions compared to Lambda expressions](../../local-functions-vs-lambdas.md).
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: Did you mean "you can use a lambda expression to implement functionality also supported by a local function"?

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+Local functions make the intent of your code clear. Anyone reading you code can see that the method is private and is not callable except by the containing method. For team projects, they also make it impossible for another developer to mistakenly call the method directly from elsewhere in the class or stuct.
+ 
+## Local function syntax
+
+A local function is defined as a nested method inside a containing member. Its definition has the following syntax:
+
+```csharp
+<return-value> <method-name> <parameter-list>
+```
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: I don't think C# syntax highlighting is appropriate for this code block, especially since it's likely going to highlight `return` as a keyword (it does that on GiHub).

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+Unlike a method definition, a local function definition cannot include the following elements:
+
+- The [private](..\..\language-reference\keywords\private.md) member access modifier. Because all local functions are private, including the `private` keyword generates compiler error CS0106, "The modifier 'private' is not valid for this item."
+ 
+- The [static](..\..\language-reference\keywords\static.md) keyword. Whether a local function is static or instance depends on its enclosing member. Including the `static` keyword generates compiler error CS0106, "The modifier 'static' is not valid for this item."
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: As I already mentioned on the CS0106 article, local functions also can't use many other modifiers. I *think* the only valid modifiers on local functions are `async` and `unsafe`, so maybe it would be easier to list those here?

Also, a local function (including its parameters and type parameters) can't have attributes, that might be worth mentioning here.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+title: "Local functions (C# Programming Guide) | Microsoft Docs"
+ms.date: "2017-06-14"
+ms.prod: .net
+ms.technology: 
+  - "devlang-csharp"
+ms.topic: "article"
+dev_langs: 
+  - "CSharp"
+helpviewer_keywords: 
+  - "fields [C#]"
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: I'm not sure what `helpviewer_keywords` are exactly, but I don't think it should be "fields" here.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+- The [private](..\..\language-reference\keywords\private.md) member access modifier. Because all local functions are private, including the `private` keyword generates compiler error CS0106, "The modifier 'private' is not valid for this item."
+ 
+- The [static](..\..\language-reference\keywords\static.md) keyword. Whether a local function is static or instance depends on its enclosing member. Including the `static` keyword generates compiler error CS0106, "The modifier 'static' is not valid for this item."
+
+In addition, note that any local variables, including any method parameters, defined in the containing member are visible in the local function. 
+ 
+The following example defines a local function named `AppendPathSeparator` that is private to a method named `GetText`:
+   
+[!code-cs[LocalFunctionExample](../../../../samples/snippets/csharp/programming-guide/classes-and-structs/local-functions1.cs)]
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: As far as I can tell, this will include the whole file. Wouldn't it be better to show just the relevant parts (no `using`s, `class` or `Main`)? I think this applies to the following snippets too.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+In addition, note that any local variables, including any method parameters, defined in the containing member are visible in the local function. 
+ 
+The following example defines a local function named `AppendPathSeparator` that is private to a method named `GetText`:
+   
+[!code-cs[LocalFunctionExample](../../../../samples/snippets/csharp/programming-guide/classes-and-structs/local-functions1.cs)]  
+   
+## Local functions and exceptions
+
+One of the useful features of local functions is that they can allow exceptions to surface immediately. For method iterators, exceptions are surfaced only when the returned sequence is enumerated, and not when the iterator is retrieved. For async methods, any exceptions thrown by a task are surfaced when the task is awaited.
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: What does "exceptions thrown by a task" mean? Did you mean "exceptions thrown by a method"?

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+   
+## Local functions and exceptions
+
+One of the useful features of local functions is that they can allow exceptions to surface immediately. For method iterators, exceptions are surfaced only when the returned sequence is enumerated, and not when the iterator is retrieved. For async methods, any exceptions thrown by a task are surfaced when the task is awaited. 
+
+The following example defines an `OddSequence` method that enumerates odd numbers between a specified range. Because it passes a number greater than 100 to the `OddSequence` enumerator method, the method throws an <xref:System.ArgumentOutOfRangeException>. As the output from the example shows, the exception surfaces only when you iterate the numbers, and not when you retrieve the enumerator.
+
+[!code-cs[LocalFunctionIterator1](../../../../samples/snippets/csharp/programming-guide/classes-and-structs/local-functions-iterator1.cs)] 
+
+Instead, you can throw an exception when performing validation and before retrieving the iterator by retrieving the iterator and returning it from a local function, as the following example shows.
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: "you can throw an exception […] before retrieving the iterator by retrieving the iterator" sounds confusing to me.

File: samples/snippets/csharp/programming-guide/classes-and-structs/local-functions-async1.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+   }
+
+   static async Task<int> GetMultipleAsync(int secondsDelay)
+   {
+      Console.WriteLine("Executing GetMultipleAsync...");
+      if (secondsDelay < 0 || secondsDelay > 5)
+         throw new ArgumentOutOfRangeException("secondsDelay cannot exceed 5.");
+         
+      await Task.Delay(secondsDelay * 1000);
+      return secondsDelay * (new Random()).Next(2,10);
```
Comment (svick) [inline samples/snippets/csharp/programming-guide/classes-and-structs/local-functions-async1.cs]: The parentheses around `new Random()` are not necessary.

File: samples/snippets/csharp/programming-guide/classes-and-structs/local-functions-async1.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+      Console.WriteLine($"The returned value is {result:N0}");
+   }
+
+   static async Task<int> GetMultipleAsync(int secondsDelay)
+   {
+      Console.WriteLine("Executing GetMultipleAsync...");
+      if (secondsDelay < 0 || secondsDelay > 5)
+         throw new ArgumentOutOfRangeException("secondsDelay cannot exceed 5.");
+         
+      await Task.Delay(secondsDelay * 1000);
```
Comment (svick) [inline samples/snippets/csharp/programming-guide/classes-and-structs/local-functions-async1.cs]: I personally prefer to use the `TimeSpan` overload to `Delay()`, especially in example code, since I think it makes the code cleaner and clearer. I.e.: `await Task.Delay(TimeSpan.FromSeconds(secondsDelay));`. But it's just a personal preference.

File: docs/csharp/language-reference/compiler-messages/cs0106.md (hunk 1 above)
Comment (BillWagner) [inline docs/csharp/language-reference/compiler-messages/cs0106.md]: I agree with removing this sentence. I think this is an implementation detail, and the behavior won't be guaranteed across versions.

File: docs/csharp/language-reference/compiler-messages/cs0106.md (hunk 2 above)
Comment (BillWagner) [inline docs/csharp/language-reference/compiler-messages/cs0106.md]: Agreed. I think the simplest explanation is that no access modifiers are allowed on local functions. (It's not clear what they would mean in the context of a local function anyway.)

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md (hunk 9 above)
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: I'd reword that last as "any exceptions thrown in an async method are observed with the returned task is awaited."

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+- Lambda expressions
+- Finalizers
+- Other local functions
+
+However, local functions can't be declared inside an expression-bodied member.
+
+> [!NOTE]
+> In some cases, you can use a lambda expression to implement functionality also supported by a lambda expression. For a comparison, see [Local functions compared to Lambda expressions](../../local-functions-vs-lambdas.md).
+
+Local functions make the intent of your code clear. Anyone reading you code can see that the method is private and is not callable except by the containing method. For team projects, they also make it impossible for another developer to mistakenly call the method directly from elsewhere in the class or stuct.
```
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: I would remove "is private and" from this sentence. It's not private per se, but it is scoped to its containing method.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md (hunk 5 above)
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: We  might try antlr syntax here, if you want.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+<return-value> <method-name> <parameter-list>
+```
+
+Unlike a method definition, a local function definition cannot include the following elements:
+
+- The [private](..\..\language-reference\keywords\private.md) member access modifier. Because all local functions are private, including the `private` keyword generates compiler error CS0106, "The modifier 'private' is not valid for this item."
+ 
+- The [static](..\..\language-reference\keywords\static.md) keyword. Whether a local function is static or instance depends on its enclosing member. Including the `static` keyword generates compiler error CS0106, "The modifier 'static' is not valid for this item."
+
+In addition, note that any local variables, including any method parameters, defined in the containing member are visible in the local function.
```
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: nit: "accessible" might be better than "visible."

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+Instead, you can throw an exception when performing validation and before retrieving the iterator by retrieving the iterator and returning it from a local function, as the following example shows.
+
+[!code-cs[LocalFunctionIterator2](../../../../samples/snippets/csharp/programming-guide/classes-and-structs/local-functions-iterator2.cs)]
+
+Local functions can be used in a similar way to handle exceptions outside of the asynchronous operation, which eliminates the need to examine the inner exceptions of an <xref:System.AggregateException>. The following example uses an asynchronous method named `GetMultipleAsync` to pause for a specified number of seconds and return a value that is a random multiple of that number of seconds. The maximum delay is 5 seconds; an <xref:System.ArgumentOutOfRangeException> results if the value is greater than 5. As the following example shows, the exception that is thrown when a value of 6 is passed to the `GetMultipleAsync` method is wrapped in an <xref:System.AggregateException> after the `GetMultipleAsync` method begins execution.
```
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: When developers call async methods, the await operator is a better way to avoid the inner exception. 

Using local functions for async work is more about failing fast, and causing the exception to be thrown synchronously, and observed synchronously.

**Review by BillWagner — CHANGES_REQUESTED** [review]
Overall, this is great and clear. 

I agree with most of the comments from @svick and added some thoughts as well.  After that, it looks great.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md (hunk 8 above)
Comment (rpetrusha) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: @svick, yes, it does include the entire file. That was intentional -- I wanted to include all of the source code so that the example could be complied and executed.

**Comment by rpetrusha** [comment]
@svick and @BillWagner, thanks for the feedback. I've addressed your comments.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+> [!NOTE]
+> In some cases, you can use a lambda expression to implement functionality also supported by a local function. For a comparison, see [Local functions compared to Lambda expressions](../../local-functions-vs-lambdas.md).
+
+Local functions make the intent of your code clear. Anyone reading you code can see that the method is not callable except by the containing method. For team projects, they also make it impossible for another developer to mistakenly call the method directly from elsewhere in the class or stuct.
+ 
+## Local function syntax
+
+A local function is defined as a nested method inside a containing member. Its definition has the following syntax:
+
+```antlr
```
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: The grammar, in antlr format, is here: https://github.com/dotnet/roslyn/blob/master/docs/features/local-functions.md

**Review by BillWagner — APPROVED** [review]
There's one change on the grammar.  After that, let's ship it.

**Comment by rpetrusha** [comment]
Thanks, @BillWagner. Could you take a look at the antlr syntax once the build completes?

**Comment by BillWagner** [comment]
@rpetrusha I like the choice of using the pseudo-code example here over the ANTLR grammar. It's more approachable to the larger audience. Those interested in seeing the grammar can use the language spec.  Let's :shipit:

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 17]
```diff
@@ -37,7 +37,29 @@ Local functions make the intent of your code clear. Anyone reading you code can
...
+   ;
+
+local-function-modifiers
+   : (async | unsafe)
+   ;
+
+local-function-body
+   : block
+   | arrow-expression-body
+   ;
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: This grammar says that a local function can be `async` or `unsafe`, but not both, which is not true: `async unsafe Task f() { }` compiles fine as a local function (even though you can't use `await` in it).

I would write the grammar like this (it's consistent with [how the spec defines modifiers for methods](https://github.com/dotnet/csharplang/blob/251cbe0/spec/classes.md#methods)):

```antlr
declaration_statement
   : local_variable_declaration ';'
   | local_constant_declaration ';'
   | local_function_declaration
   ;

local_function_declaration
   : local_function_header local_function_body
   ;

local_function_header
   : local_function_modifier* return_type identifier type_parameter_list?
        '(' formal_parameter_list? ')' type_parameter_constraints_clauses
   ;

local_function_modifier
   : 'async'
   | 'unsafe'
   ;

local_function_body
   : block
   | arrow_expression_body
   ;
```

I have also fixed few formal issues with the grammar (missing single quotes, underscores instead of hyphens).

Or should I take this to the Roslyn repo and report back here once these changes are accepted there?

Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: Oh, you changed it back. Nevermind then.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 18]
```diff
@@ -36,30 +36,8 @@ Local functions make the intent of your code clear. Anyone reading you code can
...
-local-function-modifiers
-   : (async | unsafe)
-   ;
-
-local-function-body
-   : block
-   | arrow-expression-body
-   ;
+```txt
+<modifiers: async | unsafe> <return-value> <method-name> <parameter-list>
```
Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: Nit: I think it should be `<return-type>`, not `<return-value>`.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md (hunk 17 above)
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: @svick Your comments are valid.  I'd direct them to the document I referenced from the csharplang repo.  /cc @gafter @madstorgersen for confirmation.

Comment (svick) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: @BillWagner I have created a PR: https://github.com/dotnet/roslyn/pull/20368.

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md (hunk 18 above)
Comment (rpetrusha) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md]: Yes, it should be, @svick. Thanks for pointing it out.
